core-java-modules/core-java-collections-list-7/src/test/java/com/baeldung/filterbyanyfield/FilterListByAnyMatchingFieldUnitTest.java

```diff
@@ -0,0 +1,188 @@
+        "Let's learn how to play a guitar.",
+        100);
+
+    // @formatter:on
+    private static final List<Book> BOOKS = List.of(JAVA, KOTLIN, PYTHON, GUITAR);
+
+    List<Book> fullTextSearchByLogicalOr(List<Book> books, String keyword) {
+        return books.stream()
+            .filter(book -> book.getTitle()
+                .contains(keyword) || book.getIntro()
+                .contains(keyword) || book.getTags()
+                .stream()
+                .anyMatch(tag -> tag.contains(keyword)))
+            .toList();
+    }
+
+    @Test
+    void whenFilteringInAllFields_thenCorrect() {
+        List<Book> byJava = fullTextSearchByLogicalOr(BOOKS, "Java");
+        assertThat(byJava).containsExactlyInAnyOrder(JAVA, KOTLIN);
+
+        List<Book> byArt = fullTextSearchByLogicalOr(BOOKS, "Art");
+        assertThat(byArt).containsExactlyInAnyOrder(JAVA, GUITAR);
+
+        List<Book> byLets = fullTextSearchByLogicalOr(BOOKS, "Let's");
+        assertThat(byLets).containsExactlyInAnyOrder(KOTLIN, GUITAR);
+    }
+
+    @Test
+    void whenCallingBookStrForFiltering_thenCorrect() {
+        String expected = """
+            Let's Dive Into Kotlin Codes
+            It is big fun learning how to write Kotlin codes.
+            Tech
+            Java
+            Kotlin""";
+        assertThat(KOTLIN.strForFiltering()).isEqualTo(expected);
+    }
+
+    List<Book> fullTextSearchByStrForFiltering(List<Book> books, String keyword) {
+        return books.stream()
+            .filter(book -> book.strForFiltering()
+                .contains(keyword))
+            .toList();
+    }
+
+    @Test
+    void whenFilteringInAllFieldsUsingStrForFiltering_thenCorrect() {
+        List<Book> byJava = fullTextSearchByStrForFiltering(BOOKS, "Java");
+        assertThat(byJava).containsExactlyInAnyOrder(JAVA, KOTLIN);
+
+        List<Book> byArt = fullTextSearchByStrForFiltering(BOOKS, "Art");
+        assertThat(byArt).containsExactlyInAnyOrder(JAVA, GUITAR);
+
+        List<Book> byLets = fullTextSearchByStrForFiltering(BOOKS, "Let's");
+        assertThat(byLets).containsExactlyInAnyOrder(KOTLIN, GUITAR);
+    }
+
+    boolean fullTextSearchOnObject(Object obj, String keyword, String... excludedFields) {
+        Field[] fields = obj.getClass()
+            .getDeclaredFields();
+        for (Field field : fields) {
+            if (Arrays.stream(excludedFields)
+                .noneMatch(exceptName -> exceptName.equals(field.getName()))) {
+                field.setAccessible(true);
+                try {
+                    Object value = field.get(obj);
+                    if (value != null && value.toString()
+                        .contains(keyword)) {
+                        return true;
+                    }
+                    if (!field.getType()
+                        .isPrimitive() && !(value instanceof String)) {
+                        return fullTextSearchOnObject(value, keyword);
+                    }
```

Comment:
This will exit early with false, I think we need something like this instead?
```suggestion
                    if (!field.getType()
                        .isPrimitive() && !(value instanceof String) && value != null && fullTextSearchOnObject(value, keyword)) {
                        return true;
                    }
```

Reply:
(not sure what kind of test case would catch this)

Reply:
Good catch 👍 . We should only return in case the recursion returns true.
I'd make it two if(...) return true blocks instead of one single if() with "or". I think it's easier to read.